README.md
[![Gem Version](https://badge.fury.io/rb/spotlite.png)](http://badge.fury.io/rb/spotlite)

# Spotlite

Spotlite is a ruby gem to retrieve movie information from IMDb.

## Installation

Add this line to your application's Gemfile:

    gem 'spotlite'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install spotlite

## Usage

    > require 'spotlite'
    # Access movie directly by its IMDb ID
    > movie = Spotlite::Movie.new('0133093')
    # Or use search instead
    > list = Spotlite::Movie.find('the matrix')
    > movie = list.first
    > movie.title
    => "The Matrix"
    > movie.runtime
    => 136
    > movie.genres
    => ["Action", "Sci-Fi"]
    > movie.countries
    => [{:code=>"us", :name=>"USA"}, {:code=>"au", :name=>"Australia"}]
    > movie.directors
    => [#<Spotlite::Person:0x007f96a092be70 @imdb_id="0905152", @name="Andy Wachowski", @url="http://www.imdb.com/name/nm0905152/", @credits_category="Directed by", @credits_text="(as The Wachowski Brothers)">, #<Spotlite::Person:0x007f96a092bda8 @imdb_id="0905154", @name="Lana Wachowski", @url="http://www.imdb.com/name/nm0905154/", @credits_category="Directed by", @credits_text="(as The Wachowski Brothers)">]
    > movie.cast[0..2]
    => [#<Spotlite::Person:0x007f96a19521a0 @imdb_id="0000206", @name="Keanu Reeves", @url="http://www.imdb.com/name/nm0000206/", @credits_category="Cast", @credits_text="Neo">, #<Spotlite::Person:0x007f96a1951c28 @imdb_id="0000401", @name="Laurence Fishburne", @url="http://www.imdb.com/name/nm0000401/", @credits_category="Cast", @credits_text="Morpheus">, #<Spotlite::Person:0x007f96a1951a70 @imdb_id="0005251", @name="Carrie-Anne Moss", @url="http://www.imdb.com/name/nm0005251/", @credits_category="Cast", @credits_text="Trinity">]

## Important notice

Movie titles will be localized if movie has an alternative title specific to your country.
Example: _The Great Gatsby_ (http://www.imdb.com/title/tt1343092) has an alternative title _Gatsby le magnifique_ `France (imdb display title)` and will be localized accordingly based on your IP address, if you reside in France. Non-localized title is still avaliable with `original_title` method.
Sorry, there is nothing I can do about it at the moment.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Write code
4. Test it (`rake` or `rake spec`)
5. Commit your changes (`git commit -am 'Add some feature'`)
6. Push to the branch (`git push origin my-new-feature`)
7. Create new Pull Request

## Testing

Spotlite uses RSpec as a test framework. So, first make sure you have it installed

    $ gem install rspec

Run the tests

    $ rake

Spotlite uses gem FakeWeb in order to stub out HTTP responses from IMDb. These
stubs are located in `spec/fixtures` directory.

Install FakeWeb gem:

    $ gem install fakeweb

If you want to make a new feature that uses data from a page which is not stubbed out yet:

    $ cd spotlite
    $ curl -isH "Accept-Language: en-us" http://www.imdb.com/title/tt[IMDB_ID]/ > spec/fixtures/tt[IMDB_ID]/index

or, for example:

    $ curl -isH "Accept-Language: en-us" http://www.imdb.com/title/tt[IMDB_ID]/fullcredits > spec/fixtures/tt[IMDB_ID]/fullcredits

You get the idea. And don't forget to add corresponding line to `IMDB_SAMPLES`
hash in `spec/spec_helper.rb` file.

Sometimes IMDb makes changes to its HTML layout. When this happens, Spotlite will not return
expected data, or more likely, methods will return nil or empty arrays.
First, run tests with `LIVE_TEST=true` environment variable:

    $ LIVE_TEST=true rake

Adjust methods that are failing, according to the new layout. And refresh fixtures:

    $ rake refresh_fixtures

It will run through all elements of `IMDB_SAMPLES` hash to get fresh data.

## License

Copyright (c) 2013-2014 Artem Pakk

MIT License

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
## v0.8.10 25-Jun-2016

* Added `@response` to `Person` object when fetching data

## v0.8.9 08-Jun-2016

* Added critic review source URL parsing

## v0.8.8 22-May-2016

* Ignore missing taglines placeholder

## v0.8.7 09-May-2016

* Added `taglines` method for `Movie` class to get an array of taglines

## v0.8.6 27-Feb-2016

* Numerous parsing fixes (title, original title, year, rating, metascore, votes, description, genres, runtime, content rating, starred actors)

## v0.8.5 22-Nov-2015

* Added `summaries` method for `Movie` class to get an array of plot summaries

## v0.8.4 17-Nov-2015

* Fixed content rating parsing

## v0.8.3 14-Dec-2014

* Spotlite now uses HTTParty as HTTP client, adding `response` attribute to Movie instance
* Fixed #parse_crew method failing when crew category was missing
* Recommended movies will now include only feature films

## v0.8.2 18-Jul-2014

* Clean some junk from credits text (particularly writers credits)

## v0.8.1 12-Jul-2014

* Added specs for `#find` and `#search` of `Movie` and `Person` classes
* Fix in `Spotlite::Movie#technical`

## v0.8.0 06-Jul-2014

* Added `Spotlite::Movie.search` class method to perform advanced search by parameters
* Added `Spotlite::Movie.find` class method for simple title search
* Added `technical` method for `Movie` class to get movie technical information
* Added `alternative_titles` method for `Movie` class
* Added `Spotlite::Person.search` class method to perform advanced search by parameters
* Added `Spotlite::Person.find` class method for simple name search
* Added `filmography` method for `Person` class
* Moved task to refresh fixtures out of namespace. Now it can be called as `rake refresh_fixtures`
* Added `rake console` task. It will automatically `require 'spotlite'`
* Removed `Spotlite::List` class and all of its subclasses


## v0.7.2 15-Dec-2013

* Fixes for updated Top 250 page
* Fixes for updated person page
* Fixes for updated movie credits page

## v0.7.1 12-Aug-2013

* Updated `images` method following IMDb mediaindex page layout changes

## v0.7.0 04-Aug-2013

* Added `images` method to `Movie` class to fetch still frames from media index page
* Added `recommended_movies` method to `Movie` class that returns a list of, well, recommended movies
* Moved license from LICENSE to README file

## v0.6.2 25-Jul-2013

* Fixed issue when movie description and storyline are cut after a link inside them
* Specified license in gemspec

## v0.6.1 24-Jul-2013

* Fixed issue with new line character in release comments

## v0.6.0 19-Jun-2013

* Added `Person` class
* Added `storyline` method to `Movie` class
* Added `content_rating` method to `Movie` class
* Fixed `release_dates` method following IMDb layout changes
* Added `comment` field to `release_dates` method output

## v0.5.0 03-May-2013

* Added parser for movies coming soon (http://www.imdb.com/movies-coming-soon/)

## v0.4.0 03-May-2013

* Added `metascore` method to `Movie` class
* Added `critic_reviews` method to `Movie` class
* Added parsers for movies opening this week, and this week's box office top 10 (http://www.imdb.com/movies-in-theaters/)
* `Spotlite::Search` now can handle empty search results
* Fixed issue with movie titles not necessarily being in the first table on search results page

## v0.3.0 28-Apr-2013

* Added IMDb Top 250 page parser
* Updated `keywords` method following movie keywords page changes on IMDb

## v0.2.1 21-Apr-2013

* Added `stars` method to get starred actors
* Refactored code
* Fixed issue when original non-english titles were returned in quotes

## v0.2.0 22-Feb-2013
* Added search
* Refreshed fixtures
* Fixed title, original title, and year methods due to IMDb layout changes

## v0.1.1 12-Feb-2013

* Fixed countries, languages, and rating methods due to IMDb layout changes
* Removed method to get MPAA content rating

## v0.1.0 26-Jan-2013

* Initial release. Movie class with methods for most relevant data


lib/spotlite/movie.rb
module Spotlite
  # Represents a movie on IMDb.com
  class Movie
    attr_accessor :imdb_id, :url, :response

    # Initialize a new movie object by its IMDb ID as a string
    #
    #   movie = Spotlite::Movie.new('0133093')
    #
    # Spotlite::Movie class objects are lazy loading. No HTTP request
    # will be performed upon object initialization. HTTP request will
    # be performed once when you use a method that needs remote data
    # Currently, all data is spread across 6 pages: main movie page,
    # /releaseinfo, /fullcredits, /keywords, /trivia, and /criticreviews
    def initialize(imdb_id, title = nil, year = nil)
      @imdb_id = "%07d" % imdb_id.to_i
      @title   = title
      @year    = year
      @url     = "http://www.imdb.com/title/tt#{@imdb_id}/"
    end

    # Returns a list of movies as an array of +Spotlite::Movie+ objects
    # Takes single parameter and searches for movies by title and alternative titles
    def self.find(query)
      results = Spotlite::Client.get(
        'http://www.imdb.com/find', query: { q: query, s: 'tt', ttype: 'ft' }
      )
      results.css('.result_text').map do |result|
        imdb_id = result.at('a')['href'].parse_imdb_id
        title   = result.at('a').text.strip
        year    = result.children.take(3).last.text.parse_year

        [imdb_id, title, year]
      end.map do |values|
        self.new(*values)
      end
    end

    # Returns a list of movies as an array of +Spotlite::Movie+ objects
    # Takes optional parameters as a hash
    # See https://github.com/defeed/spotlite/wiki/Advanced-movie-search for details
    def self.search(params = {})
      defaults = {
        title_type: 'feature',
        view: 'simple',
        count: 250,
        start: 1,
        sort: 'moviemeter,asc'
      }
      params = defaults.merge(params)
      results = Spotlite::Client.get(
        'http://www.imdb.com/search/title', query: params
      )
      results.css('td.title').map do |result|
        imdb_id = result.at('a')['href'].parse_imdb_id
        title   = result.at('a').text.strip
        year    = result.at('.year_type').text.parse_year

        [imdb_id, title, year]
      end.map do |values|
        self.new(*values)
      end
    end

    # Returns title as a string
    def title
      # strip some non-breaking space at the end
      @title ||= details.at("h1[itemprop='name']").children.first.text.strip.gsub(' ', '') rescue nil
    end

    # Returns original non-english title as a string
    def original_title
      details.at("div.originalTitle").children.first.text.gsub('"', '').strip rescue nil
    end

    # Returns year of original release as an integer
    def year
      @year ||= details.at("h1[itemprop='name'] span#titleYear a").text.parse_year rescue nil
    end

    # Returns IMDb rating as a float
    def rating
      details.at("div.imdbRating span[itemprop='ratingValue']").text.to_f rescue nil
    end

    # Returns Metascore rating as an integer
    def metascore
      details.at("div.titleReviewBar a[href^=criticreviews] span").text.to_i rescue nil
    end

    # Returns number of votes as an integer
    def votes
      details.at("div.imdbRating span[itemprop='ratingCount']").text.gsub(/[^\d+]/, '').to_i rescue nil
    end

    # Returns short description as a string
    def description
      desc = details.at("div.summary_text[itemprop='description']").text.strip.clean_description rescue nil
      (desc.nil? || desc.empty?) ? nil : desc
    end

    # Returns storyline as a string. Often is the same as description
    def storyline
      details.at("#titleStoryLine div[itemprop='description'] p").text.strip.clean_description rescue nil
    end

    # Returns a list of plot summaries as an array of strings
    def summaries
      plot_summaries.css("p.plotSummary").map { |summary| summary.text.strip }
    end

    # Returns content rating as a string
    def content_rating
      details.at("div.subtext meta[itemprop='contentRating']")['content'] rescue nil
    end

    # Returns a list of genres as an array of strings
    def genres
      details.css("div.subtext a[href^='/genre/']").map { |genre| genre.text } rescue []
    end

    # Returns a list of countries as an array of hashes
    # with keys: +code+ (string) and +name+ (string)
    def countries
      array = []
      details.css("div.txt-block a[href^='/country/']").each do |node|
        array << {:code => node['href'].clean_href, :name => node.text.strip}
      end

      array
    end

    # Returns a list of languages as an array of hashes
    # with keys: +code+ (string) and +name+ (string)
    def languages
      array = []
      details.css("div.txt-block a[href^='/language/']").each do |node|
        array << {:code => node['href'].clean_href, :name => node.text.strip}
      end

      array
    end

    # Returns runtime (length) in minutes as an integer
    def runtime
      details.at("div.subtext time[itemprop='duration']")['datetime'].gsub(/[^\d+]/, '').to_i rescue nil
    end

    # Returns primary poster URL as a string
    def poster_url
      src = details.at('div.poster img')['src'] rescue nil

      if src =~ /^(http:.+@@)/ || src =~ /^(http:.+?)\.[^\/]+$/
        $1 + '.jpg'
      end
    end

    # Returns an array of recommended movies as an array of initialized objects of +Movie+ class
    def recommended_movies
      details.css('.rec-title').reject do |node|
        # reject movies that don't have a release year yet
        node.at('span').nil?
      end.reject do |node|
        # reject everything other than featured film
        /\d{4}-\d{4}/.match(node.at('span').text) ||
          /Series|Episode|Video|Documentary|Movie|Special|Short|Game|Unknown/.match(node.at('span').text)
      end.map do |node|
        imdb_id = node.at("a[href^='/title/tt']")['href'].parse_imdb_id
        title   = node.at('a').text.strip
        year    = node.at('span').text.parse_year

        [imdb_id, title, year]
      end.map do |values|
        Spotlite::Movie.new(*values)
      end
    end

    # Returns a list of keywords as an array of strings
    def keywords
      plot_keywords.css("a[href^='/keyword/']").map { |keyword| keyword.text.strip } rescue []
    end

    # Returns a list of trivia facts as an array of strings
    def trivia
      movie_trivia.css("div.sodatext").map { |node| node.text.strip } rescue []
    end

    # Returns a list of taglines as an array of strings
    def taglines
      movie_taglines.css("#taglines_content > .soda").map { |node| node.text.clean_tagline }
    end

    # Returns a list of movie alternative titles as an array of hashes
    # with keys +title+ (string) and +comment+ (string)
    def alternative_titles
      array = []
      release_info.css('#akas').css('tr').map do |row|
        cells = row.css('td')
        array << { :title => cells.last.text.strip, :comment => cells.first.text.strip }
      end

      array
    end

    # Returns a list of directors as an array of +Spotlite::Person+ objects
    def directors
      parse_crew('Directed by')
    end

    # Returns a list of writers as an array of +Spotlite::Person+ objects
    def writers
      parse_crew('Writing Credits')
    end

    # Returns a list of producers as an array of +Spotlite::Person+ objects
    def producers
      parse_crew('Produced by')
    end

    # Returns a list of starred actors as an array of +Spotlite::Person+ objects
    def stars
      details.css("div.plot_summary_wrapper span[itemprop='actors'] a[href^='/name/nm']").map do |node|
        imdb_id = node['href'].parse_imdb_id
        name = node.text.strip

        [imdb_id, name]
      end.map do |values|
        Spotlite::Person.new(*values)
      end
    end

    # Returns a list of actors as an array +Spotlite::Person+ objects
    def cast
      full_credits.css('table.cast_list tr').reject do |row|
        # Skip 'Rest of cast' row
        row.children.size == 1
      end.map do |row|
        imdb_id = row.at('td:nth-child(2) a')['href'].parse_imdb_id
        name = row.at('td:nth-child(2) a').text.strip_whitespace
        credits_text = row.last_element_child.text.strip_whitespace

        [imdb_id, name, 'Cast', credits_text]
      end.map do |values|
        Spotlite::Person.new(*values)
      end
    end

    # Returns a list of crew members of a certain category as an array +Spotlite::Person+ objects
    def parse_crew(category)
      table = full_credits.search("[text()^='#{category}']").first.next_element rescue nil
      if table && table.name == 'table'
        table.css('tr').reject do |row|
          # Skip empty table rows with one non-braking space
          row.text.strip.size == 1
        end.map do |row|
          imdb_id = row.first_element_child.at('a')['href'].parse_imdb_id
          name = row.first_element_child.at('a').text.strip_whitespace
          credits_text = row.last_element_child.text.strip_whitespace.clean_credits_text

          [imdb_id, name, category, credits_text]
        end.map do |values|
          Spotlite::Person.new(*values)
        end
      else
        []
      end
    end

    # Combines all crew categories and returns an array of +Spotlite::Person+ objects
    def crew
      crew_categories.map{ |category| parse_crew(category) }.flatten
    end

    # Returns combined `cast` and `crew` as an array of +Spotlite::Person+ objects
    def credits
      cast + crew
    end

    # Returns available crew categories, e.g. "Art Department", "Writing Credits", or "Stunts", as an array of strings
    def crew_categories
      array = []
      full_credits.css('h4.dataHeaderWithBorder').reject{ |h| h['id'] == 'cast' }.map do |node|
        array << (node.children.size > 1 ? node.children.first.text.strip_whitespace : node.children.text.strip_whitespace)
      end

      array
    end

    # Returns a list of regions and corresponding release dates
    # as an array of hashes with keys:
    # region +code+ (string), +region+ name (string), +date+ (date), and +comment+ (string)
    # If day is unknown, 1st day of month is assigned
    # If day and month are unknown, 1st of January is assigned
    def release_dates
      array = []
      table = release_info.at('#release_dates')
      table.css('tr').map do |row|
        cells = row.css('td')
        code = cells.first.at('a')['href'].clean_href.split('=').last.downcase rescue nil
        region = cells.first.at('a').text rescue nil
        date = cells.at('.release_date').text.strip.parse_date
        comment = cells.last.text.strip.clean_release_comment
        comment = nil if comment.empty?

        array << { :code => code, :region => region, :date => date, :comment => comment }
      end unless table.nil?

      array
    end

    # Returns original release date as a date
    def release_date
      release_dates.first[:date] rescue nil
    end

    # Returns a list of critic reviews as an array of hashes
    # with keys: +source+ (string), +author+ (string), +excerpt+ (string), and +score+ (integer)
    def critic_reviews
      array = []
      reviews.css("tr[itemprop='reviews']").map do |review|
        source = review.at("b[itemprop='publisher'] span[itemprop='name']").text
        author = review.at("span[itemprop='author'] span[itemprop='name']").text
        url = review.at("a[itemprop='url']")['href'] rescue nil
        excerpt = review.at("div[itemprop='reviewbody']").text.strip
        score = review.at("span[itemprop='ratingValue']").text.to_i

        array << {
          source: source.empty? ? nil : source,
          author: author.empty? ? nil : author,
          url: url,
          excerpt: excerpt,
          score: score
        }
      end

      array
    end

    # Returns URLs of movie still frames as an array of strings
    def images
      array = []
      still_frames.css('#media_index_thumbnail_grid img').map do |image|
        src = image['src'] rescue nil

        if src =~ /^(http:.+@@)/ || src =~ /^(http:.+?)\.[^\/]+$/
          array << $1 + '.jpg'
        end
      end

      array
    end

    # Returns technical information like film length, aspect ratio, cameras, etc. as a hash of arrays of strings
    def technical
      hash = {}
      table = technical_info.at_css('#technical_content table') rescue nil

      table.css('tr').map do |row|
        hash[row.css('td').first.text.strip] = row.css('td').last.children.
          map(&:text).
          map(&:strip_whitespace).
          reject(&:empty?).
          reject{|i| i == '|'}.
          slice_before{|i| /^[^\(]/.match i}.
          map{|i| i.join(' ')}
      end unless table.nil?

      hash
    end

    private

    def details # :nodoc:
      @details ||= open_page
    end

    def release_info # :nodoc:
      @release_info ||= open_page('releaseinfo')
    end

    def full_credits # :nodoc:
      @full_credits ||= open_page('fullcredits')
    end

    def plot_keywords # :nodoc:
      @plot_keywords ||= open_page('keywords')
    end

    def movie_trivia # :nodoc:
      @movie_trivia ||= open_page('trivia')
    end

    def reviews
      @reviews ||= open_page('criticreviews')
    end

    def still_frames # :nodoc:
      @still_frames ||= open_page('mediaindex', {refine: 'still_frame'})
    end

    def technical_info # :nodoc:
      @technical_info ||= open_page('technical')
    end

    def plot_summaries # :nodoc:
      @plot_summaries ||= open_page('plotsummary')
    end

    def movie_taglines # :nodoc:
      @taglines ||= open_page('taglines')
    end

    def open_page(page = nil, query = {}) # :nodoc:
      response = Spotlite::Client.get "#{@url}#{page}", query: query
      @response = { code: response.code, message: response.message } and response
    end
  end

end
